api/genericresource/resource_management.go

```diff
@@ -0,0 +1,196 @@
+	taskReservations []*api.GenericResource) error {
+	resSelected := make([]*api.GenericResource, 0)
+
+	for _, res := range taskReservations {
+		tr := res.GetDiscrete()
+		if tr == nil {
+			return newError("task should only hold Discrete type")
+		}
+
+		// Select the resources
+		nrs, err := selectNodeResources(nodeAvailableResources, tr)
+		if err != nil {
+			return err
+		}
+
+		resSelected = append(resSelected, nrs...)
+	}
+
+	ClaimResources(nodeAvailableResources, taskAssigned, resSelected)
+	return nil
+}
+
+// ClaimResources adds the specified resources to the task's list
+// and removes them from the node's generic resource list
+func ClaimResources(nodeAvailableResources, taskAssigned *[]*api.GenericResource,
+	resSelected []*api.GenericResource) {
+	*taskAssigned = append(*taskAssigned, resSelected...)
+	RemoveNodeResources(nodeAvailableResources, resSelected)
+}
+
+func selectNodeResources(nodeRes *[]*api.GenericResource,
+	tr *api.GenericDiscrete) (nrs []*api.GenericResource, err error) {
+	nrs = make([]*api.GenericResource, 0, tr.Value)
+
+loop:
+	for _, res := range *nodeRes {
+		if getKind(res) != tr.Kind {
+			continue
+		}
+
+		switch nr := res.Resource.(type) {
+		case *api.GenericResource_Discrete:
+			if nr.Discrete.Value >= tr.Value && tr.Value != 0 {
+				nrs = append(nrs, NewDiscrete(tr.Kind, tr.Value))
+			}
+
+			break loop
+		case *api.GenericResource_Str:
+			nrs = append(nrs, res)
+
+			if int64(len(nrs)) == tr.Value {
+				break loop
+			}
+		}
+	}
+
+	if len(nrs) == 0 {
+		return nil, newError("Not enough resource available for task reservations: %+v\n", tr)
+	}
+
+	return nrs, nil
+}
+
+// Reclaim adds the resources taken by the task to the node's store
+func Reclaim(nodeAvailableResources *[]*api.GenericResource, taskAssigned []*api.GenericResource) error {
+	// The node could have been updated
+	if nodeAvailableResources == nil {
+		return newError("Node no longer has any resources")
+	}
+
+	for _, res := range taskAssigned {
+		switch tr := res.Resource.(type) {
+		case *api.GenericResource_Discrete:
+			nrs := GetResource(tr.Discrete.Kind, *nodeAvailableResources)
+
+			// If the resource went down to 0 it's no longer in the
+			// available list
+			if len(nrs) == 0 {
+				*nodeAvailableResources = append(*nodeAvailableResources, res)
+			}
+
+			if len(nrs) != 1 {
+				continue // Type change
+			}
+
+			nr := nrs[0].GetDiscrete()
+			if nr == nil {
+				continue // Type change
+			}
+
+			nr.Value += tr.Discrete.Value
+		case *api.GenericResource_Str:
+			*nodeAvailableResources = append(*nodeAvailableResources, res)
+		}
+	}
+
+	return nil
+}
+
+// Sanitize checks that nodeAvailableResources does not add resources unknown
+// to the nodeSpec (nodeRes) or goes over the integer bound specified
+// by the spec.
+// Note this is because the user is able to update a node's resources
+func Sanitize(nodeRes []*api.GenericResource, nodeAvailableResources *[]*api.GenericResource) {
```

Comment:
I like the idea of building this functionality into `Reclaim` instead of making it a separate function. Unless I am missing something, it looks like `Sanitize` and `Reclaim` are only ever called as a pair. It would make the scheduler code simpler if it only has to call one function to get this correct.